core/src/main/java/io/questdb/griffin/engine/functions/groupby/CountDistinctIntGroupByFunction.java

```diff
@@ -0,0 +1,123 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.engine.functions.GroupByFunction;
+import io.questdb.griffin.engine.functions.LongFunction;
+import io.questdb.std.IntHashSet;
+import io.questdb.std.Numbers;
+import io.questdb.std.ObjList;
+
+public class CountDistinctIntGroupByFunction extends LongFunction implements GroupByFunction {
+    private final Function arg;
+    private final ObjList<IntHashSet> sets = new ObjList<>();
+    private int valueIndex;
+    private int setIndex;
+
+    public CountDistinctIntGroupByFunction(Function arg) {
+        this.arg = arg;
+    }
+
+    @Override
+    public void computeFirst(MapValue mapValue, Record record) {
+        final IntHashSet set;
+        if (sets.size() <= setIndex) {
+            sets.extendAndSet(setIndex, set = new IntHashSet());
+        } else {
+            set = sets.getQuick(setIndex);
+        }
+
+        set.clear();
+        final int val = arg.getInt(record);
+        if (val != Numbers.INT_NaN) {
+            set.add(val);
+            mapValue.putLong(valueIndex, 1L);
+        } else {
+            mapValue.putLong(valueIndex, 0L);
+        }
+        mapValue.putInt(valueIndex + 1, setIndex++);
+    }
+
+    @Override
+    public void computeNext(MapValue mapValue, Record record) {
+        final IntHashSet set = sets.getQuick(mapValue.getInt(valueIndex + 1));
+        final int val = arg.getInt(record);
+        if (val != Numbers.INT_NaN) {
+            final int index = set.keyIndex(val);
+            if (index < 0) {
+                return;
+            }
+            set.addAt(index, val);
+            mapValue.addLong(valueIndex, 1);
+        }
+    }
+
+    @Override
+    public void pushValueTypes(ArrayColumnTypes columnTypes) {
+        this.valueIndex = columnTypes.getColumnCount();
+        columnTypes.add(ColumnType.LONG);
+        columnTypes.add(ColumnType.INT);
+    }
+
+    @Override
+    public void setEmpty(MapValue mapValue) {
+        mapValue.putLong(valueIndex, 0L);
+    }
+
+    @Override
+    public void setLong(MapValue mapValue, long value) {
+        mapValue.putLong(valueIndex, value);
+    }
+
+    @Override
+    public void setNull(MapValue mapValue) {
+        mapValue.putLong(valueIndex, Numbers.LONG_NaN);
+    }
+
+    @Override
+    public long getLong(Record rec) {
+        return rec.getLong(valueIndex);
+    }
+
+    @Override
+    public boolean isConstant() {
+        return false;
+    }
+
+    @Override
+    public boolean isReadThreadSafe() {
+        return true;
```

Comment:
it has some state, why is it thread safe?

Reply:
Probably, that's because aggregate functions don't participate in filter expressions. So, this flag changes nothing, but we have it set in `Count*` functions for some reasons.

Removed the overrides in d0ad60ad3a685655b993c30fd5e955c7cafb9138